src/applications/Applications.DicomLoader/Loader.cs

```diff
@@ -0,0 +1,165 @@
+        using var archive=ReaderFactory.Open(archiveStream);
+        while(archive.MoveToNextEntry())
+        {
+            if (archive.Entry.IsDirectory) continue;
+            using var entry = archive.OpenEntryStream();
+            var ds = DicomFile.Open(entry).Dataset;
+            Process(ds,$"{fi.FullName}!{archive.Entry.Key}",fi.DirectoryName??throw new ApplicationException($"No parent directory for {fi.FullName}"),archive.Entry.Size,iStore,ct);
+        }
+    }
+
+    private static void Process(DicomDataset ds, string path, string directoryName, long size,
+        IMongoCollection<BsonDocument> iStore, CancellationToken ct)
+    {
+        // Consider flushing every 256 file loads
+        if ((Interlocked.Increment(ref _fileCount) & 0xff) == 0)
+        {
+            Flush();
+            Report();
+        }
+
+        var identifiers = new string[3];
+
+        // Pre-fetch these to ensure they exist before we go further
+        identifiers[0] = ds.GetValue<string>(DicomTag.StudyInstanceUID, 0);
+        identifiers[1] = ds.GetValue<string>(DicomTag.SeriesInstanceUID, 0);
+        identifiers[2] = ds.GetValue<string>(DicomTag.SOPInstanceUID, 0);
+
+        if (identifiers.Any(string.IsNullOrWhiteSpace))
+        {
+            Console.WriteLine($"'{path}' had blank DICOM UID");
+            return;
+        }
+
+        DicomFileMessage message = new()
+        {
+            StudyInstanceUID = identifiers[0],
+            SeriesInstanceUID = identifiers[1],
+            SOPInstanceUID = identifiers[2],
+
+            DicomFileSize = size,
+            DicomFilePath = path
+        };
+        // ReSharper disable once InconsistentlySynchronizedField
+        _seriesList.AddOrUpdate(identifiers[1],id=>LoadSm(id,directoryName,ds,identifiers[0]) , IncSm);
+        DicomDataset filtered = new(ds.Where(i => i is not DicomOtherByteFragment).ToArray());
+
+        iStore.InsertOne(
+            new BsonDocument("header", MongoDocumentHeaders.ImageDocumentHeader(message, new MessageHeader())).AddRange(
+                DicomTypeTranslaterReader.BuildBsonDocument(filtered)), cancellationToken: ct);
+    }
+
+    public static ValueTask Load(string filename, CancellationToken ct)
+    {
+        if (ImageStore is null) throw new ArgumentNullException(nameof(ImageStore));
+        if (SeriesStore is null) throw new ArgumentNullException(nameof(SeriesStore));
+        if (Database is null) throw new ArgumentNullException(nameof(Database));
+
+        if (!File.Exists(filename))
+        {
+            Console.WriteLine($@"{filename} does not exist, skipping");
+            return ValueTask.CompletedTask;
+        }
+
+        if (ImageStore.CountDocuments(
+                new BsonDocumentFilterDefinition<BsonDocument>(new BsonDocument("header", new BsonDocument("DicomFilePath", filename))),new CountOptions(),ct) > 0)
+        {
+            Console.WriteLine($@"{filename} already loaded, skipping");
```

Comment:
Thinking about this... we are ignoring duplicates.  Should we instead be booting the old copy and putting in the newer one? that way if we change the implementation or something or find a subset are corrupt it is easier to just rerun stuff?

Reply:
That might be a useful option in future, my aim was to have it re-runnable and efficiently load only any files which had been missed/added. I'll add something like a --reload option to overwrite existing data in a later version.